tests/src/common/test_case.rs

```diff
@@ -98,33 +91,87 @@ where
             profit_addr: None,
             leaser_addr: None,
             lpp_addr: None,
-            oracle: None,
-            timealarms: None,
+            oracle_addr: None,
+            time_alarms_addr: None,
             lease_code_id: None,
-            _lpn: PhantomData,
         }
     }
 
-    pub fn send_funds(&mut self, user_addr: &Addr, funds: Vec<CwCoin>) -> &mut Self {
+    pub fn send_funds_from_admin(&mut self, user_addr: Addr, funds: &[CwCoin]) -> &mut Self {
         self.app
-            .send_tokens(Addr::unchecked(ADMIN), user_addr.clone(), &funds)
+            .send_tokens(Addr::unchecked(ADMIN), user_addr, funds)
             .unwrap();
 
         self
     }
 
-    pub fn init(&mut self, user: &Addr, mut init_funds: Vec<CwCoin>) -> &mut Self {
+    pub fn dispatcher(&self) -> &Addr {
+        self.dispatcher_addr.as_ref().unwrap()
+    }
+
+    pub fn treasury(&self) -> &Addr {
+        self.treasury_addr.as_ref().unwrap()
+    }
+
+    pub fn profit(&self) -> &Addr {
+        self.profit_addr.as_ref().unwrap()
+    }
+
+    pub fn leaser(&self) -> &Addr {
+        self.leaser_addr.as_ref().unwrap()
+    }
+
+    pub fn lpp(&self) -> &Addr {
+        self.lpp_addr.as_ref().unwrap()
+    }
+
+    pub fn oracle(&self) -> &Addr {
+        self.oracle_addr.as_ref().unwrap()
+    }
+
+    pub fn time_alarms(&self) -> &Addr {
+        self.time_alarms_addr.as_ref().unwrap()
+    }
+
+    pub fn lease_code_id(&self) -> u64 {
+        self.lease_code_id.unwrap()
+    }
+}
+
+pub struct TestCase<Lpn> {
+    test_case: GenericTestCase,
+    _lpn: PhantomData<Lpn>,
+}
+
+impl<Lpn> TestCase<Lpn>
+where
+    Lpn: Currency,
+{
+    pub const LEASER_CONNECTION_ID: &'static str = "connection-0";
+
+    pub fn new() -> Self {
+        Self::with_reserve(&[cwcoin::<Lpn, _>(10_000)])
+    }
+
+    pub fn with_reserve(reserve: &[CwCoin]) -> Self {
+        Self {
+            test_case: GenericTestCase::with_reserve(reserve),
+            _lpn: PhantomData,
+        }
+    }
+
+    pub fn init(&mut self, user: Addr, init_funds: &mut [CwCoin]) -> &mut Self {
```

Comment:
please merge this fn into the `new` and even `with_reserve`. The goal is to have the fn that creates the instance initialized with the minimum set of data.